### DISCUSSION TITLE
Problems Monitoring 2 interfaces from separate networks using bond0

### DISCUSSION THREAD

**Discussion by gh0st-network** [Security-Onion-Solutions/securityonion#2676] [Unsupported Versions] [answered]
Hi,

I'm trying to have SO 2 monitoring interfaces capture packets from two (2) separate networks (192.168.x.x and 172.16.x.x) via two (2) separate network interfaces (identified as ens224 and ens256 respectively). Understand in SO2, the monitoring interfaces are combined into a single bonded interface (bond0).

Based on the documentation. discussions, and setup wizard, it appears to be a straight-forward process where both interfaces are selected as monitored interfaces during the setup process.

Upon testing, I realized that bond0 did not function as it was supposed to. I was only able to capture packets from the first network (192.168.x.x) and there were zero (0) packets from the second network (172.16.x.x). The ifconfig output and bond0 details are listed below. Also, it appears that the ifconfig output for both ens224 and ens256 interfaces have the same hardware address which is weird (google search shows a few such occurrences where the solution provided was to view the bond0 details - included below).

To verify both TAPs were configured correctly, an instance of SO 16.04 was deployed to monitor on the same 172.16.x.x interface on the same VM host and it correctly captured packets. However, the same interface on the SO2 instance did not see a single packet in tcpdump.

Appreciate if anyone is able to provide any pointers, suggestions or fixes.

Thank you.

```
bond0: flags=1475<UP,BROADCAST,RUNNING,NOARP,PROMISC,MASTER>  mtu 1500
        ether 00:50:56:3a:bc:de  txqueuelen 1000  (Ethernet)
        RX packets 1487664  bytes 226546165 (226.5 MB)
        RX errors 0  dropped 181  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

docker0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 172.17.0.1  netmask 255.255.255.0  broadcast 172.17.0.255
        ether 02:42:68:be:e9:04  txqueuelen 0  (Ethernet)
        RX packets 72833034  bytes 264208229184 (264.2 GB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 82794822  bytes 267176220751 (267.1 GB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

ens160: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.1.25  netmask 255.255.255.0  broadcast 192.168.1.255
        ether 00:50:56:32:af:a3  txqueuelen 1000  (Ethernet)
        RX packets 4244587  bytes 1253769350 (1.2 GB)
        RX errors 0  dropped 777622  overruns 0  frame 0
        TX packets 1131299  bytes 437733343 (437.7 MB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

ens192: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 172.16.8.25  netmask 255.255.255.0  broadcast 172.16.8.255
        ether 00:50:56:3a:ff:ff  txqueuelen 1000  (Ethernet)
        RX packets 2279175  bytes 325138742 (325.1 MB)
        RX errors 0  dropped 16  overruns 0  frame 0
        TX packets 929767  bytes 94412495 (94.4 MB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

ens224: flags=2499<UP,BROADCAST,RUNNING,NOARP,PROMISC,SLAVE>  mtu 1500
        ether 00:50:56:3a:bc:de  txqueuelen 1000  (Ethernet)
        RX packets 1487664  bytes 226546165 (226.5 MB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

ens256: flags=2499<UP,BROADCAST,RUNNING,NOARP,PROMISC,SLAVE>  mtu 1500
        ether 00:50:56:3a:bc:de  txqueuelen 1000  (Ethernet)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        loop  txqueuelen 1000  (Local Loopback)
        RX packets 17735260  bytes 22906501369 (22.9 GB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 17735260  bytes 22906501369 (22.9 GB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0


Ethernet Channel Bonding Driver: v3.7.1 (April 27, 2011)

Bonding Mode: load balancing (round-robin)
MII Status: up
MII Polling Interval (ms): 100
Up Delay (ms): 0
Down Delay (ms): 0

Slave Interface: ens224
MII Status: up
Speed: 10000 Mbps
Duplex: full
Link Failure Count: 0
Permanent HW addr: 00:50:56:3a:bc:de
Slave queue ID: 0

Slave Interface: ens256
MII Status: up
Speed: 10000 Mbps
Duplex: full
Link Failure Count: 0
Permanent HW addr: 00:50:56:3a:ff:0e
```

**Comment by TOoSmOotH** [comment]
Are you using Ubuntu?

**Reply by gh0st-network** [reply]
Yes. Ubuntu.

```
Distributor ID: Ubuntu
Description:    Ubuntu 18.04.5 LTS
Release:        18.04
Codename:       bionic
```

**Reply by weslambert** [reply]
What is the output of `nmcli con show`?

**Reply by gh0st-network** [reply]
```
NAME                UUID                                  TYPE      DEVICE
bond0               0e99b376-6615-4449-98a6-dd9855ec7044  bond      bond0
bond0-slave-ens224  644ddca1-95f6-4e3a-a923-01f4374b56b9  ethernet  ens224
bond0-slave-ens256  77249781-8600-4043-b936-6e536ff7480c  ethernet  ens256
docker0             5471fa0f-248a-4a0b-8056-5994db2b469b  bridge    docker0
```

**Comment by gh0st-network** [comment]
Hi everyone,  to quickly provide an update, I have further tested the setup and noted the following:

**ens256 (172.16.x.x)** is only problematic when its part of **bond0**.

Removing the interface from the bond using `sudo nmcli con down "bond0-slave-ens256"` results in packets being seen by the OS.

```
NAME                UUID                                  TYPE      DEVICE
bond0               0e99b376-6615-4449-98a6-dd9855ec7044  bond      bond0
bond0-slave-ens224  644ddca1-95f6-4e3a-a923-01f4374b56b9  ethernet  ens224
docker0             5471fa0f-248a-4a0b-8056-5994db2b469b  bridge    docker0
bond0-slave-ens256  77249781-8600-4043-b936-6e536ff7480c  ethernet  --
```

ifconfig output

```
ens256: flags=451<UP,BROADCAST,RUNNING,NOARP,PROMISC>  mtu 1500
        ether 00:50:56:3a:ff:0e  txqueuelen 1000  (Ethernet)
        RX packets 182  bytes 23213 (23.2 KB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

(a few seconds later)

ens256: flags=451<UP,BROADCAST,RUNNING,NOARP,PROMISC>  mtu 1500
        ether 00:50:56:3a:ff:0e  txqueuelen 1000  (Ethernet)
        RX packets 5693  bytes 6359383 (6.3 MB)
        RX errors 0  dropped 1  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

```
The packets stop being seen once the command `sudo nmcli con up "bond0-slave-ens256"` is issued.

So far,
`sudo tcpdump -i ens256` sees packets only when its disconnected from bond0
`sudo tcpdump -i ens224` sees packets regardless connection status on bond0

Appreciate if anyone can share if this is a known issue or bug. Its looking very weird for me.

Thanks in advance.

**Comment by rsharris3** [comment]
I was able to get a setup like this working, in the /etc/sysconfig/network-scripts/ifcfg-bond0 file change the bond mode from the default round-robin (balance-rr) to Adaptive load balancing (balance-alb).

```
BONDING_OPTS=mode=balance-alb
TYPE=Bond
BONDING_MASTER=yes
HWADDR=
MTU=1500
PROXY_METHOD=none
BROWSER_ONLY=no
IPV6INIT=no
NAME=bond0
UUID=
DEVICE=bond0
ONBOOT=yes
```

**Reply by ghost** [reply]
Can confirm that this works. 

Another solution is via root cronjobs that set interface MAC ADDRs to their Permanent HW ADDR after sleeping:
```
# Gather Permanent HW ADDRs
cat /proc/net/bonding/bond0

# Update the interface ID and MACs, then set in root user crontab
@reboot sleep 60 && /sbin/ip link set eth0 address 00:00:00:00:05:01
@reboot sleep 60 && /sbin/ip link set eth1 address 00:00:00:00:05:02
@reboot sleep 60 && /sbin/ip link set eth2 address 00:00:00:00:05:03
```


Had issues on VMware Fusion Pro/Workstation Pro/ESXi. The source of the issue seems to be that VMware will only route SPAN traffic to interfaces that are hard assigned the appropriate MAC address when `Mac Address Changed` is set to Reject.


This is expected behavior per the kernel docs:
- Kernel Docs on bonding: https://www.kernel.org/doc/Documentation/networking/bonding.txt
- Prettier version here: https://wiki.linuxfoundation.org/networking/bonding

```
8.  Where does a bonding device get its MAC address from?

  When using slave devices that have fixed MAC addresses, or when
the fail_over_mac option is enabled, the bonding device's MAC address is
the MAC address of the active slave.

  For other configurations, if not explicitly configured (with
ifconfig or ip link), the MAC address of the bonding device is taken from
its first slave device.  This MAC address is then passed to all following
slaves and remains persistent (even if the first slave is removed) until
the bonding device is brought down or reconfigured.

  If you wish to change the MAC address, you can set it with
ifconfig or ip link:

# ifconfig bond0 hw ether 00:11:22:33:44:55

# ip link set bond0 address 66:77:88:99:aa:bb

  The MAC address can be also changed by bringing down/up the
device and then changing its slaves (or their order):

# ifconfig bond0 down ; modprobe -r bonding
# ifconfig bond0 .... up
# ifenslave bond0 eth...

  This method will automatically take the address from the next
slave that is added.

  To restore your slaves' MAC addresses, you need to detach them
from the bond (`ifenslave -d bond0 eth0'). The bonding driver will
then restore the MAC addresses that the slaves had before they were
enslaved.
```


https://docs.vmware.com/en/VMware-vSphere/7.0/com.vmware.vsphere.security.doc/GUID-942BD3AA-731B-4A05-8196-66F2B4BF1ACB.html#GUID-942BD3AA-731B-4A05-8196-66F2B4BF1ACB

> When the Mac address changes option is set to Reject, ESXi does not honor requests to change the effective MAC address to a different address than the initial MAC address. This setting protects the host against MAC impersonation. The port that the virtual machine adapter used to send the request is disabled and the virtual machine adapter does not receive any more frames until the effective MAC address matches the initial MAC address. The guest operating system does not detect that the MAC address change request was not honored.

I'm assuming this and not making a declaration:
Any time that the BOND interface is configured and the slaves are joined, the interfaces likely broadcast a heads up to VMWare letting it know that the Effective HW ADDR (MAC Address) has changed from the Permanent HW ADDR to mirror the first interface indexed by the kernel. When this happens, ESXis port security blacklists the interface to block possible mac spoofing attack.

In short:

1. Run `cat /proc/net/bonding/bond0`
2. The first *Slave Interface* listed has the Effective HW ADDR for all other slave interfaces.
3. Run `ip link` and inspect the interfaces to confirm that their Effective HW ADDRs mirror the first indexed slave interface.

---

Really, it makes me curious on the choice of using a bond interface for aggregating traffic. Does any one know if there's a storyboard for the decision to shift to bonded interfaces?

**Reply by gh0st-network** [reply]
Hi all, thanks for the suggestions and solutions. Wanted to update that I was able to resolve the issue but forgot to update the post. Solution I used was to enable "Allow MAC Changes" within the VM Host settings (this is under the Networking Settings). Ensure "Allow MAC Changes" is applied at both the vSwitch and Port Group levels.

The above was the only setting I needed to apply to get the packet capture to work. Hope this helps someone experiencing this issue.

**Reply by darthkamal** [reply]
Thank you @gh0st-network, this fixed the issue for me as well.

**Reply by dougburks** [reply]
I've added this to https://docs.securityonion.net/en/2.3/vmware.html#esxi.

**Comment by dougburks** [comment, marked as answer]
Documenting @gh0st-network's answer here so it can be marked as answered:
> Hi all, thanks for the suggestions and solutions. Wanted to update that I was able to resolve the issue but forgot to update the post. Solution I used was to enable "Allow MAC Changes" within the VM Host settings (this is under the Networking Settings). Ensure "Allow MAC Changes" is applied at both the vSwitch and Port Group levels.
The above was the only setting I needed to apply to get the packet capture to work. Hope this helps someone experiencing this issue.

I've added this to https://docs.securityonion.net/en/2.3/vmware.html#esxi.
